libdrgn/dwarf_info.c

```diff
@@ -9069,62 +9069,81 @@ drgn_oil_type_iterator_create(struct drgn_program *prog,
-					return drgn_error_format(DRGN_ERROR_LOOKUP, "DWARF DIE with address 0x%lx has no children", (uintptr_t) class_die.addr);
-				bool template_parameter_found = false, get_object_size_func_found = false, has_sibling = true;
-				while (has_sibling && !(template_parameter_found && get_object_size_func_found)) {
-					if (dwarf_tag(&child) == DW_TAG_template_type_parameter) {
-						template_parameter = child;
-						template_parameter_found = true;
-					} else if (dwarf_tag(&child) == DW_TAG_subprogram) {
-						const char* child_name = dwarf_diename(&child);
-						if (child_name && strcmp(child_name, "getObjectSize") == 0) {
-							Dwarf_Attribute linkage_name_attr;
-							if (dwarf_attr_integrate(&child, DW_AT_linkage_name, &linkage_name_attr)) {
-								struct drgn_symbol *symbol;
-								// We need to check that the symbol exists here and not at the end of the loop because
-								// getObjectSize is overloaded in CodegenHandler: it's possible that not all of the various
-								// overloads are used, so we need to check that the one we're currently examining *is* used
-								// (by verifying that it exists in the symbol table).
-								err = drgn_program_find_symbol_by_name(iter->prog, dwarf_formstring(&linkage_name_attr), &symbol);
-								if (!err) {
-									*function_addr_ret = drgn_symbol_address(symbol);
-									drgn_symbol_destroy(symbol);
-									get_object_size_func_found = true;
-								} else {
-									drgn_error_destroy(err);
-								}
-							}
-						}
-					}
-					has_sibling = dwarf_siblingof(&child, &child) == 0;
+		if (!dwarf_attr_integrate(&class_die, DW_AT_name, &name_attr))
+			continue;
+
+		const char* name = dwarf_formstring(&name_attr);
+		if (!name || strncmp(name, CODEGEN_HANDLER_PREFIX, CODEGEN_HANDLER_PREFIX_LENGTH) != 0)
+			continue;
+
+		Dwarf_Die child, template_parameter;
+		if (dwarf_child(&class_die, &child) != 0)
+			return drgn_error_format(DRGN_ERROR_LOOKUP,
+					"DWARF DIE with address 0x%lx has no children",
+					(uintptr_t) class_die.addr);
+
+		bool template_parameter_found = false, get_object_size_func_found = false;
+		for (bool has_sibling = true;
+				 has_sibling && !(template_parameter_found && get_object_size_func_found);
+				 has_sibling = dwarf_siblingof(&child, &child) == 0) {
+			if (dwarf_tag(&child) == DW_TAG_template_type_parameter) {
+				template_parameter = child;
+				template_parameter_found = true;
+			} else if (dwarf_tag(&child) == DW_TAG_subprogram) {
+				const char* child_name = dwarf_diename(&child);
+				if (!child_name || strcmp(child_name, "getObjectSize") != 0)
+					continue;
+
+				Dwarf_Attribute linkage_name_attr;
+				if (!dwarf_attr_integrate(&child, DW_AT_linkage_name, &linkage_name_attr))
+					continue;
+
+				// Only return this function if it exists in the symbol table. Some
+				// overloads may have been removed by compiler optimisations.
+				struct drgn_symbol *symbol;
+				err = drgn_program_find_symbol_by_name(iter->prog,
```

Comment:
It dawned on me that we could run into some issues if `getObjectSize` is inlined (which is possible if the binary using OIL is compiled with LTO). I need to change the way `drgn_program_find_symbol_by_name` handles inlined functions in conjunction with some of my other work on inlining, will remember to make sure this case is addressed.